feat(share-account): implement details screen

markrizkalla opened the pull request:
### Summary
Implemented the **Share Account Details** screen, enabling users to view account summaries, transaction history, and charges. This implementation aligns the Share Account feature with the existing Loan and Savings account architecture.

### Changes

**Data Layer**
* **Updated** `ShareAccountRepository` and `ShareAccountService` to fetch account details with associations (`purchasedShares`, `charges`).
* **Updated** `ClientChargeRepository` to support fetching charges for Share Accounts.
* **Added** core models: `ShareAccountWithAssociations`, `Summary`, `Transactions`, `Currency`, and `EnumOptionData`.

**UI**
* **Details Screen:** Created `ShareAccountDetailsScreen` displaying a dashboard grid (Approved/Pending shares, Market Price, Status) and Action Cards.
* **Navigation:** Added `shareNavGraph` and updated `AuthenticatedNavigation` to route users to the new details screen when clicking a Share Account.
* **Actions:** Implemented `ShareActionItems` to handle navigation for **Charges**, **Transactions**, and **QR Code**.

**Logic & Integration**
* **Transactions:** Updated `AccountsTransactionViewModel` to fetch Share transactions and implemented a mapper to convert share-specific `purchasedShares` into the generic `Transaction` model.
* **Charges:** Updated `ClientChargeViewModel` to handle `ChargeType.SHARE`, allowing reuse of the existing Charges screen.
* **Color Logic:** Implemented specific logic in the transaction mapper to display "Charge Payments" and "Purchases" as **Debits (Red)**.

**Fixes**
* Fixed a potential crash in `ChargeDetailsRoute` by handling empty `dueDate` values (displaying `"-"` instead).

### Fixes
Fixes - [MM-455](https://mifosforge.jira.com/jira/software/c/projects/MM/boards/65?selectedIssue=MM-455)

### Screenshots

| Before | After |
|:---:|:---:|
| Feature does not exist | **Details Dashboard** <br> <img src="https://github.com/user-attachments/assets/6f2f03c6-3b0d-4b7f-9bdb-d39f9b6f9c46" width="300" alt="Details" /> |
| Feature does not exist | **Transaction History** <br> <img src="https://github.com/user-attachments/assets/cde41225-d3a7-4b79-9431-7cf8299e54bb" width="300" alt="Transactions" /> |
| Feature does not exist | **Charges List** <br> <img src="https://github.com/user-attachments/assets/cdd34891-8b3a-4f31-8bc8-8e531e3694d5" width="300" alt="Charges" /> |

### Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the static analysis check `./gradlew check` or `ci-prepush.sh` to make sure you didn't break anything
- [x] If you have multiple commits please combine them into one commit by squashing them.

[MM-455]: https://mifosforge.jira.com/browse/MM-455?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * View Share Account Details: summary, status, market price, approved/pending shares, key dates and actions (Charges, Transactions, QR).
  * Browse Share Transactions and Share Charges with currency-aware formatting, grouping and QR support.
* **Integrations**
  * Share-account flows added to authenticated navigation and transaction/charges screens.
* **Improvements**
  * Distinguish network vs server errors; better retry behavior.
* **Localization**
  * Added multilingual strings including a generic server-error message.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

markrizkalla committed:
feat(share-account): implement details screen

- Implement Share Account Details dashboard with summary grid (approved/pending shares, market price).
- Add navigation Action Cards for Charges, Transactions, and QR Code.
- Implement data mapper in RecentTransactionRepository to adapt share-specific purchasedShares to the generic Transaction model.
- Update AccountsTransactionViewModel to fetch and display Share transactions.
- Integrate standard Charge model for Share Accounts in ClientChargeViewModel.
- Fix crash caused by missing dueDate in activation charges.
- Correct credit/debit logic to ensure 'Charge Payment' is displayed as a debit.

feat(share-account): implement details screen

- Implement Share Account Details dashboard with summary grid (approved/pending shares, market price).
- Add navigation Action Cards for Charges, Transactions, and QR Code.
- Implement data mapper in RecentTransactionRepository to adapt share-specific purchasedShares to the generic Transaction model.
- Update AccountsTransactionViewModel to fetch and display Share transactions.
- Integrate standard Charge model for Share Accounts in ClientChargeViewModel.
- Fix crash caused by missing dueDate in activation charges.
- Correct credit/debit logic to ensure 'Charge Payment' is displayed as a debit.

markrizkalla committed:
Fix(share-account): Update Share Account models and UI

This commit introduces several fixes and improvements to the share account feature:

-   Updates `ShareAccountWithAssociations`, `Summary`, and `Transactions` models to handle `null` or empty values gracefully and align data types (`savingsAccountNumber`, `accountNo`, `purchasedDate`).
-   Enhances network error detection in `ShareAccountDetailsViewModel` by checking the exception's cause.
-   Sorts the action items in `ShareAccountDetailsScreen` for a consistent and ordered display.
-   Clears the charges list in `ClientChargeViewModel` when an empty success state is received.

markrizkalla committed:
Refactor(share-account): Make ShareAccountWithAssociations fields nullable

Made fields in the `ShareAccountWithAssociations` data class nullable to prevent crashes from null API responses. This change ensures greater stability by gracefully handling missing data.

Adjusted the `ShareAccountDetailsViewModel` and `TransactionViewModel` to safely handle these new nullable fields, providing default values where necessary. Also performed minor code cleanup in the share account feature.

markrizkalla committed:
Refactor(share-account): Make ShareAccountWithAssociations fields nullable

Made fields in the `ShareAccountWithAssociations` data class nullable to prevent crashes from null API responses. This change ensures greater stability by gracefully handling missing data.

Adjusted the `ShareAccountDetailsViewModel` and `TransactionViewModel` to safely handle these new nullable fields, providing default values where necessary. Also performed minor code cleanup in the share account feature.

biplab1 commented on core/model/src/commonMain/kotlin/org/mifos/mobile/core/model/entity/accounts/share/Summary.kt:
```diff
@@ -0,0 +1,28 @@
+    val timeline: Timeline? = null,
+    val totalApprovedShares: Int? = null,
+    val totalPendingForApprovalShares: Int? = null,
+) : Parcelable
```
`@Parcelize` can be removed as shown below. Please remove it from everywhere it is present in this PR.
```suggestion
data class Summary(
    val id: Long? = null,
    val accountNo: String? = null,
    val productId: Long? = null,
    val productName: String? = null,
    val status: Status? = null,
    val currency: Currency? = null,
    val timeline: Timeline? = null,
    val totalApprovedShares: Int? = null,
    val totalPendingForApprovalShares: Int? = null,
)
```

biplab1 commented on core/model/src/commonMain/kotlin/org/mifos/mobile/core/model/entity/accounts/share/EnumOptionData.kt:
```diff
@@ -0,0 +1,22 @@
+
+@Serializable
+@Parcelize
+data class EnumOptionData(
```
Use an appropriate name for the data class (also for the file) and update accordingly wherever the class has been used.

biplab1 commented on feature/client-charge/src/commonMain/kotlin/org/mifos/mobile/feature/charge/charges/ClientChargeViewModel.kt:
```diff
@@ -22,6 +22,7 @@ import mifos_mobile.feature.client_charge.generated.resources.client_charges
 import mifos_mobile.feature.client_charge.generated.resources.feature_generic_error_server
 import mifos_mobile.feature.client_charge.generated.resources.loan_charges
```
Name of string resource used in a feature module (say, checker-inbox-name), should be `feature_[module name]_[string reference name]`. For example, `feature_checker_inbox_task_[string reference name]`.

Please follow the above naming convention wherever required.

markrizkalla replied:
client-charge module already have so many string resources and in so many other languages, would not it be better to make this refactor in another PR ?

biplab1 reviewed (changes requested):
Please make sure the @coderabbitai suggestions have been implemented. 

Suggested some minor changes. Kindly upload the screen recording showing navigation.

markrizkalla committed:
Refactor(share-account): Remove Parcelize and fix various bugs

- Remove Parcelable and Parcelize from core data models as they are no longer needed.
- Prefix `share_charges` string resource with `feature_client_charge_` for consistency.
- Add null check in ShareAccountDetailsViewModel to prevent crashes.
- Improve IOException handling in TransactionViewModel.

markrizkalla committed:
Refactor(share-account): Cleanup imports and formatting

markrizkalla committed:
Merge branch 'development' into feature/share-account-details-screen

markrizkalla committed:
Merge remote-tracking branch 'origin/feature/share-account-details-screen' into feature/share-account-details-screen

markrizkalla committed:
Refactor(share-account): Standardize string resource keys

Renamed string resources within the `feature/share-account` module to follow a consistent naming convention, prefixing them with `feature_share_account_`. This improves clarity and organization.

Additionally, added a defensive check in `TransactionViewModel` to prevent crashes by providing a default date if `purchasedDate` is incomplete.

markrizkalla committed:
Merge remote-tracking branch 'upstream/development' into feature/share-account-details-screen

# Conflicts:
#	feature/accounts/src/commonMain/kotlin/org/mifos/mobile/feature/accounts/accountTransactions/TransactionViewModel.kt

markrizkalla committed:
Refactor(share-account): Handle potential null `totalPendingForApprovalShares`

markrizkalla committed:
Merge branch 'development' into feature/share-account-details-screen

biplab1 commented on core/data/src/commonMain/kotlin/org/mifos/mobile/core/data/repositoryImpl/ClientChargeRepositoryImp.kt:
```diff
@@ -62,4 +62,16 @@ class ClientChargeRepositoryImp(
+    override fun getShareAccountCharges(shareAccountId: Long): Flow<DataState<List<Charge>>> {
+        return dataManager.shareAccountApi.getShareAccountDetails(shareAccountId)
+            .map { response ->
+                // response.charges is already List<Charge>
```
Please avoid these kinds of comments unless very important.

biplab1 commented on feature/accounts/src/commonMain/kotlin/org/mifos/mobile/feature/accounts/accountTransactions/TransactionViewModel.kt:
```diff
@@ -297,10 +306,108 @@ internal class AccountsTransactionViewModel(
+            }
+            DataState.Loading -> updateState { it.copy(uiState = ScreenUiState.Loading) }
+            is DataState.Success -> {
+                // MAP ShareTransactions -> UiTransaction
```
Same here, please review.

biplab1 commented on feature/accounts/src/commonMain/kotlin/org/mifos/mobile/feature/accounts/accountTransactions/TransactionViewModel.kt:
```diff
@@ -297,10 +306,108 @@ internal class AccountsTransactionViewModel(
+        isCredit = when {
+            type?.value?.contains("Purchase", ignoreCase = true) == true -> false // Red (Debit)
+            type?.value?.contains("Charge Payment", ignoreCase = true) == true -> false // Red (Debit)
+            else -> true // Green (Credit)
```
Please avoid inline comments

biplab1 commented on feature/share-account/src/commonMain/kotlin/org/mifos/mobile/feature/shareaccount/shareAccountDetails/ShareAccountDetailsScreen.kt:
```diff
@@ -0,0 +1,219 @@
+                        isActive = state.isActive,
+                    )
+
+                    // (Bottom Section)
```
Please remove these kinds of comments wherever present.

biplab1 reviewed (approved):
Some minor changes requested. Rest looks good to me. This can be merged after the updates have been implemented.

markrizkalla committed:
Refactor(share-account): Clean up comments and code in account features
